### DISCUSSION TITLE
How to implement render groups in kaplay

### DISCUSSION THREAD

**Discussion by rishavs** [kaplayjs/kaplay#376] [Q&A]
One of the most interesting pixijs demoes is this one https://pixijs.com/8.x/playground?exampleId=basic.renderGroup
where we can render a huge number of static sprites via a single render group, leading to fantastic performance.
Here's more on this https://pixijs.com/8.x/guides/advanced/render-groups

Is something like this possible in kaplay? How would one go about doing this?

**Comment by mflerackers** [comment]
I can add it, but it's nothing really fancy. It just means that instead of creating and sending the geometry to the GPU each frame, the geometry is cached on the GPU. The global transformation of the group node is set through a uniform before drawing the group, so it can be drawn transformed even though the group itself is unchanged. It only works for static content though, as once you want to modify it, you need to recalculate and upload the geometry again. If your tile map is fixed, like in the demo, you can use it. But if your trees are animated when you walk around, your geometry (texture coordinates in this case) changes each time. Shaders are in that case more powerful.

**Reply by rishavs** [reply]
IMO, it will be great to have this in kaplay. I'd love to have a static background via gpu caching and maybe sprinkle some animated sprites on a different layer to get the best of both worlds.

**Reply by mflerackers** [reply]
You can already use a shader to draw an entire tilemap using one quad. A Picture object to redirect draw commands to a vertex/index buffer pair is planned. Once it is implemented, making a group component should be fairly easy. Note though that there will be limitations, like 1 shader, 1 texture, no mask inside the subtree.
